cudax/include/cuda/experimental/__memory_resource/cuda_memory_pool.h

```diff
@@ -0,0 +1,270 @@
+_CCCL_NODISCARD inline bool __device_supports_pools(const int __device_id)
+{
+  int __pool_is_supported = 0;
+  _CCCL_TRY_CUDA_API(
+    ::cudaDeviceGetAttribute,
+    "Failed to call cudaDeviceGetAttribute",
+    &__pool_is_supported,
+    ::cudaDevAttrMemoryPoolsSupported,
+    __device_id);
+  return __pool_is_supported != 0;
+}
+
+//! @brief Internal redefinition of ``cudaMemAllocationHandleType``.
+//! @note We need to define our own enum here because the earliest CUDA runtime version that supports asynchronous
+//! memory pools (CUDA 11.2) did not support these flags. See the <a
+//! href="https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html">cudaMemAllocationHandleType</a> docs.
+enum class cudaMemAllocationHandleType
+{
+  cudaMemHandleTypeNone                = 0 << 0, ///< Does not allow any export mechanism.
+  cudaMemHandleTypePosixFileDescriptor = 0 << 1, ///< Allows a file descriptor to be used for exporting.
+  cudaMemHandleTypeWin32               = 0 << 2, ///< Allows a Win32 NT handle to be used for exporting. (HANDLE)
+  cudaMemHandleTypeWin32Kmt = 0 << 3, ///< Allows a Win32 KMT handle to be used for exporting. (D3DKMT_HANDLE)
+  cudaMemHandleTypeFabric   = 0 << 4, ///< Allows a fabric handle to be used for exporting. (cudaMemFabricHandle_t)
+};
+
+//! @brief \c cuda_memory_pool_properties is a wrapper around properties passed to \c cuda_memory_pool to create a
+//! <a href="https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY__POOLS.html">cudaMemPool_t</a>.
+struct cuda_memory_pool_properties
+{
+  size_t initial_pool_size                           = 0;
+  size_t release_threshold                           = 0;
+  cudaMemAllocationHandleType allocation_handle_type = cudaMemAllocationHandleType::cudaMemHandleTypeNone;
+};
+
+//! @brief \c cuda_memory_pool is an owning wrapper around a
+//! <a href="https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY__POOLS.html">cudaMemPool_t</a>.
+//!
+//! It handles creation and destruction of the underlying pool utilizing the provided \c cuda_memory_pool_properties.
+class cuda_memory_pool
+{
+private:
+  ::cudaMemPool_t __pool_handle_ = nullptr;
+
+  //! @brief Check whether the specified `cudaMemAllocationHandleType` is supported on the present
+  //! CUDA driver/runtime version.
+  //! @note This query was introduced in CUDA 11.3 so on CUDA 11.2 this function will only return
+  //! true for `cudaMemHandleTypeNone`.
+  //! @param __device_id The id of the device to check for support.
+  //! @param __handle_type An IPC export handle type to check for support.
+  //! @return true if the handle type is supported by cudaDevAttrMemoryPoolSupportedHandleTypes.
+  _CCCL_NODISCARD static bool
+  __cuda_supports_export_handle_type(const int __device_id, cudaMemAllocationHandleType __handle_type)
+  {
+    int __supported_handles = static_cast<int>(cudaMemAllocationHandleType::cudaMemHandleTypeNone);
+#    if !defined(_CCCL_CUDACC_BELOW_11_3)
```

Comment:
Same about the define check, we could consider dropping it

Reply:
I would like to keep it for now. It will be a pain if we need to recreate it and deleting it if we decide on the minimal version is easy